### PR TITLE
Issue #1410 and #1267 - display expiration date and update domain creation date

### LINKED ISSUE #1410: Users can view expiration date for their domain(s)
### Issue description

Once a domain is registered and DNS is set up, the primary interaction most users have had with the .gov registrar is to renew their domain. However, we don't show the expiration date to users yet.

Allow domain managers to view expiration date for their domain(s) per the current design.

[Figma design link](https://www.figma.com/file/aJbuDLJNNikqQObTuCNKQa/GetGov_Designs_ECS-Truss?type=design&node-id=2392%3A87560&mode=design&t=D8bl5kO5czg2RpbF-1)

### Acceptance criteria
Check Figma link above for look and functionality of these items
- [ ] On the dashboard, remove Date created column for both Domains and Domain requests
- [ ] On Domains, add column showing Expiry date
- [ ] On Domain requests, add Date submitted column
- [ ] When a domain request is in Started status, it should show "Not submitted" in place of a submission date. This text is in disabled color. (See Figma)
- [ ] On Overview pages add two lines below Status alert. These show the expiry date and date created.
- [ ] On Expired domains only, on the Overview page, add (expired) in red at the end of the Expires line. (See Figma)

### Additional context

DO NOT MERGE until #1212 has been executed (i.e., the expiration dates have been updated in the system, so that the user does not see OLD data)

### Links to other issues

🔄 Corresponding design ticket #1213 
⛔ Blocked by #1212

### PR DESCRIPTION
**Pull request by dave-kennedy-ecs**
## Ticket

Resolves #1410 #1267 

## Changes

- Creation_date in domain is updated from registry.  It cannot be set otherwise.
- Added is_expired logic in domain
- Refactored _fetch_cache in domain
- Refactored index.py to utilize domain model instead of a dict (in order to use is_expired logic in home page)
- Submission_date added to domain application.
- Submission date set to today on domain application submission
- On home page, domains, Column "Expires" replaces creation date for domains
- On home page, domains, Expired is added to State column (if domain is expired)
- On home page, domain requests, Column "Date submitted" replaces creation date for domain requests
- On home page, domain requests, Not submitted (with disabled decoration) displays when domain request does not have a submission date
- On domain detail page, expiration date and creation date display
- On domain detail page, if expiration date is in the past, "expired" is also displayed in red
- On domain detail page, if expiration date is in the past, "Expired" displays in the status area at the top of page

## Context for reviewers

Design Review:
Testing submission date for domain requests - Note that test domain requests created by fixtures will not have submission date, and will display 'Not submitted' for submission date.  The submission date will only be created for new domain requests, when they are submitted.  To test this, create a new domain request, fill it out and submit it.  You should then see a submission date.
Testing Expired visibility for domains - Note that test domains without corresponding EPP domain may not have expiration date.  In this case, these domains will display as Expired.  This use case will not exist in production.  For domains with an expiration date, the expiration date will display in the Expires column.  If the expiration date is in the past, the status will display as Expired.  If you click to manage a domain, and the domain is expired, the status will display as Expired.  Below the status, Expiration date and Creation date will display.

Some screenshots are shown below for expected behavior

Dev Review:
migrations - note this does have a migrations file as the domain application model now has a submission date
domain.py - is_expired is added; _fetch_cache is refactored (to satisfy linter); creation_date updated through _fetch_cache
domain_application.py - submission_date is added; it is updated through submit method
domain_information.py - accounting for submission_date at time of domain_application submission and creation of domaininfo
templates - functionality described above
tests - account for changes to expiration_date, creation_date and submission_date throughout
views/index.py - refactored to expose domain model to template in order to use is_expired from domain model in template; refactor is also for simplicity

#### Satisfied acceptance criteria and met development standards

- [ ] Met the acceptance criteria, or will meet them in a subsequent PR
- [ ] Created/modified automated tests
- [ ] Added at least 2 developers as PR reviewers (only 1 will need to approve)
- [ ] Messaged on Slack or in standup to notify the team that a PR is ready for review
- [ ] Changes to “how we do things” are documented in READMEs and or onboarding guide
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the assoicated migrations file has been commited.

#### Ensured code standards are met (Original Developer)

- [ ] All new functions and methods are commented using plain language
- [ ] Did dependency updates in Pipfile also get changed in requirements.txt?
- [ ] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values

#### Validated user-facing changes (if applicable)

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)
- [ ] Add at least 1 designer as PR reviewer

#### Reviewed, tested, and left feedback about the changes

- [ ] Pulled this branch locally and tested it
- [ ] Reviewed this code and left comments
- [ ] Checked that all code is adequately covered by tests
- [ ] Made it clear which comments need to be addressed before this work is merged
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the assoicated migrations file has been commited.

#### Ensured code standards are met (Code reviewer)

- [ ] All new functions and methods are commented using plain language
- [ ] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values
- [ ] (Rarely needed) Did dependency updates in Pipfile also get changed in requirements.txt?

#### Validated user-facing changes as a developer

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Meets all designs and user flows provided by design/product
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers, the suggestion is to use ones that the developer didn't (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

**Note:** Multiple code reviewers can share the checklists above, a second reviewers should not make a duplicate checklist

#### Verified that the changes match the design intention

- [ ] Checked that the design translated visually
- [ ] Checked behavior
- [ ] Checked different states (empty, one, some, error)
- [ ] Checked for landmarks, page heading structure, and links
- [ ] Tried to break the intended flow

#### Validated user-facing changes as a designer

- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

## Screenshots

Home page showing both Domains and Domain requests
image

Domain management page for Expired domain:
image

### CODE REVIEW

**Comment by gabydisarli** [comment]
One visual design thing:

The alignment of the columns from domain to requests is off. Turns out it's aways been a little bit off. But now it's a whole lot off and it looks a little sloppy. Can we more closely align these things?

here's the tables before
Screenshot 2023-12-14 at 1 02 31 PM

And the tables on your sandbox
Screenshot 2023-12-14 at 1 03 48 PM

**Comment by gabydisarli** [comment]
Also I'm not sure if this happened on your PR but I'm noticing a bunch of empty groups read out on VO when reviewing the overview page. If this isn't part of your PR let me know I can make a ticket.

**Comment by rachidatecs** [comment]
> One visual design thing:
> 
> The alignment of the columns from domain to requests is off. Turns out it's aways been a little bit off. But now it's a whole lot off and it looks a little sloppy. Can we more closely align these things?
> 
> here's the tables before Screenshot 2023-12-14 at 1 02 31 PM
> 
> And the tables on your sandbox Screenshot 2023-12-14 at 1 03 48 PM

This makes me nervous, @gabydisarli. The basic way to align cells on 2 different tables is to artificially set a min-width on cells in a column. That's not a viable solution with dynamic content.

A more solid solution is to query in JS at runtime the width of all the cells in the 1st column in the 2 tables, then add through JS that min-width. This seems like too much code to do this, and it seems like it's not something that's "the web way." I wonder if I can convince you to let this one go? Or else let's create a ticket and spend time to do it as best as we can.

**Comment by gabydisarli** [comment]
For anyone else observing, Rachid and I chatted, and we agreed I would make a hot fix ticket to address the alignment issues.

**Review by gabydisarli — APPROVED** [review]

**Comment by rachidatecs** [comment]
> For anyone else observing, Rachid and I chatted, and we agreed I would make a hot fix ticket to address the alignment issues.

Yes, setting a fixed width will work after all as prototyping showed.

File: src/registrar/models/domain.py [hunk 1]
```diff
@@ -200,6 +201,14 @@ def creation_date(self) -> date:
         """Get the `cr_date` element from the registry."""
         return self._get_property("cr_date")
 
+    @creation_date.setter  # type: ignore
+    def creation_date(self, ex_date: date):
```
Comment (zandercymatics) [inline src/registrar/models/domain.py:205]:
```suggestion
    def creation_date(self, cr_date: date):
```

File: src/registrar/models/domain.py [hunk 2]
```diff
@@ -1582,45 +1601,19 @@ def _delete_hosts_if_not_used(self, hostsToDelete: list[str]):
...
-                    cleaned["contacts"] = old_cache_contacts
-
-            # if expiration date from registry does not match what is in db,
-            # update the db
-            if "ex_date" in cleaned and cleaned["ex_date"] != self.expiration_date:
-                self.expiration_date = cleaned["ex_date"]
-                self.save()
+            cleaned = self._clean_cache(cache, data_response)
+            self._update_hosts_and_contacts(cleaned, fetch_hosts, fetch_contacts)
+            self._update_dates(cleaned)
```
Comment (zandercymatics) [inline src/registrar/models/domain.py:1608]: Love this refactor

File: src/registrar/models/domain.py [hunk 3]
```diff
@@ -1648,6 +1650,42 @@ def _get_dnssec_data(self, response_extensions):
                 dnssec_data = extension
         return dnssec_data
 
+    def _update_hosts_and_contacts(self, cleaned, fetch_hosts, fetch_contacts):
+        """Capture and store old hosts and contacts from cache if the don't exist"""
```
Comment (zandercymatics) [inline src/registrar/models/domain.py:1654]:
```suggestion
        """Capture and store old hosts and contacts from cache if they don't exist"""
```
(minor nitpick)

File: src/registrar/templates/domain_detail.html [hunk 4]
```diff
@@ -6,7 +6,7 @@
   <div class="margin-top-4 tablet:grid-col-10">
 
     <div
-        class="usa-summary-box dotgov-status-box margin-top-3 padding-left-2{% if domain.state == domain.State.UNKNOWN or domain.state == domain.State.DNS_NEEDED%} dotgov-status-box--action-need{% endif %}"
+        class="usa-summary-box dotgov-status-box margin-top-3 padding-left-2{% if not domain.is_expired %}{% if domain.state == domain.State.UNKNOWN or domain.state == domain.State.DNS_NEEDED %} dotgov-status-box--action-need{% endif %}{% endif %}"
```
Comment (zandercymatics) [inline src/registrar/templates/domain_detail.html:9]:
```suggestion
        class="usa-summary-box dotgov-status-box margin-top-3 padding-left-2{% if not domain.is_expired and (domain.state == domain.State.UNKNOWN or domain.state == domain.State.DNS_NEEDED) %} dotgov-status-box--action-need{% endif %}"
```
(Weird spacing in the editor - don't directly commit that)
(Q) Why not include this in the original if statement?

**Review by zandercymatics — APPROVED** [review]
LGTM. Great work.

I did notice one edge case (designer input) for expired domains that haven't loaded that data from EPP yet (image below). This more or less means that the expiration date will be blank for those which the user hasn't accessed prior.

Personally, I think that hits on another issue best handled in another ticket/PR (caching work) - so I wouldn't include that work here

image

File: src/registrar/templates/domain_detail.html (hunk 4 above)
Comment (dave-kennedy-ecs) [reply src/registrar/templates/domain_detail.html:9]: This doesn't work in django templates.  Have to use nested ifs.

File: src/registrar/models/domain.py (hunk 1 above)
Comment (dave-kennedy-ecs) [reply src/registrar/models/domain.py:205]: good catch

File: src/registrar/models/domain.py (hunk 3 above)
Comment (dave-kennedy-ecs) [reply src/registrar/models/domain.py:1654]: good catch

**Pull request merged**